File: src/TarkovTextureConverter/TarkovTextureConverter.Cli/GltfUtils.cs
using Microsoft.Extensions.Logging;
using System.Text.Json;
using System.Text.Json.Nodes;

namespace TarkovTextureConverter.Cli;

public static class GltfUtils
{
    public static async Task UpdateGltfFilesAsync(string inputFolder, string outputFolderAbsPath, ILogger logger, CancellationToken cancellationToken = default)
    {
        if (!Directory.Exists(inputFolder))
        {
            logger.LogError("GLTF Update: Input folder '{InputFolder}' not found.", inputFolder);
            return;
        }

        string outputFolderBasename = Path.GetFileName(outputFolderAbsPath);
        //logger.LogInformation("Scanning for GLTF files in '{InputFolder}' to update URIs relative to '{OutputFolderBasename}'...", inputFolder, outputFolderBasename);

        int gltfFoundCount = 0;
        int updatedGltfCount = 0;

        try
        {
            // Exclude files that already end with "_converted.gltf" to prevent processing our own output
            var gltfFiles = Directory.EnumerateFiles(inputFolder, "*.gltf", SearchOption.TopDirectoryOnly)
                .Where(f => !Path.GetFileNameWithoutExtension(f).EndsWith("_converted"))
                .ToList();
            
            gltfFoundCount = gltfFiles.Count;
            
            await Parallel.ForEachAsync(gltfFiles, 
                new ParallelOptions { MaxDegreeOfParallelism = Environment.ProcessorCount, CancellationToken = cancellationToken },
                async (filePath, token) => 
                {
                    bool updated = await ProcessGltfFileAsync(filePath, outputFolderBasename, logger, token);
                    if (updated) Interlocked.Increment(ref updatedGltfCount);
                });
        }
        catch (OperationCanceledException)
        {
             logger.LogWarning("GLTF file scanning/processing was cancelled.");
        }
        catch (Exception ex) when (ex is IOException or UnauthorizedAccessException)
        {
            logger.LogError(ex, "GLTF Update: Error scanning directory {InputFolder}", inputFolder);
        }
        catch (Exception ex)
        {
            logger.LogError(ex, "GLTF Update: Unexpected error during scan");
        }

        if (gltfFoundCount == 0)
        {
            logger.LogInformation("GLTF Update: No .gltf files found in {InputFolder}.", inputFolder);
        }
        else
        {
            //logger.LogInformation("GLTF Update: Finished processing {GltfFoundCount} GLTF files. Updated {UpdatedGltfCount} files.", gltfFoundCount, updatedGltfCount);
        }
    }

    private static async Task<bool> ProcessGltfFileAsync(string filePath, string outputFolderBasename, ILogger logger, CancellationToken cancellationToken)
    {
        //logger.LogInformation("Processing GLTF file: {FilePath}", filePath);
        bool fileUpdated = false;

        try
        {
            string jsonContent = await File.ReadAllTextAsync(filePath, cancellationToken);

            JsonNode? rootNode = JsonNode.Parse(jsonContent);
            if (rootNode == null)
            {
                logger.LogError("  Failed to parse GLTF file: {FilePath}", filePath);
                return false;
            }

            JsonArray? images = rootNode["images"]?.AsArray();
            JsonArray? textures = rootNode["textures"]?.AsArray();
            JsonArray? materials = rootNode["materials"]?.AsArray();

            if (images == null || textures == null)
            {
                logger.LogInformation("  Skipping GLTF file: No 'images' or 'textures' array found.");
                return false;
            }

            var updatedImageUriMap = new Dictionary<int, string>();

            for (int i = 0; i < images.Count; i++)
            {
                JsonObject? img = images[i]?.AsObject();
                if (img == null || !img.TryGetPropertyValue("uri", out JsonNode? uriNode) || uriNode == null) continue;

                string? oldUri = uriNode.GetValue<string>();
                if (string.IsNullOrEmpty(oldUri) || oldUri.StartsWith("data:")) continue;

                string oldFilename = Path.GetFileName(oldUri.Replace("\\", "/"));
                TextureType? texType = TextureProcessor.GetTextureType(oldFilename, tarkinMode: true);

                string? newUriBase = texType switch
                {
                    TextureType.Diffuse => Utils.InsertSuffix(oldFilename, "_color"),
                    TextureType.SpecGlos => Utils.InsertSuffix(oldFilename, "_roughness"),
                    TextureType.Normal => Utils.InsertSuffix(oldFilename, "_converted"),
                    _ => null
                };

                if (!string.IsNullOrEmpty(newUriBase))
                {
                    newUriBase = Path.ChangeExtension(newUriBase, ".png");
                    string newUri = Path.Combine(outputFolderBasename, newUriBase).Replace("\\", "/");

                    if (img["uri"]?.GetValue<string>() != newUri)
                    {
                        //logger.LogInformation("    Image [{ImageIndex}]: Updating URI '{OldUri}' -> '{NewUri}'", i, oldUri, newUri);
                        img["uri"] = newUri;
                        updatedImageUriMap[i] = newUri;
                        fileUpdated = true;
                    }
                }
            }

            bool specGlossExtensionRemoved = false;
            if (fileUpdated && materials != null)
            {
                for(int matIdx = 0; matIdx < materials.Count; matIdx++)
                {
                    JsonObject? mat = materials[matIdx]?.AsObject();
                    if (mat == null) continue;
                    
                    string matName = mat["name"]?.GetValue<string>() ?? $"Material_{matIdx}";
                    bool matModified = false;

                    if (mat.TryGetPropertyValue("extensions", out JsonNode? extNode) &&
                        extNode is JsonObject extensions &&
                        extensions.TryGetPropertyValue("KHR_materials_pbrSpecularGlossiness", out JsonNode? specGlossNode) &&
                        specGlossNode is JsonObject specGloss)
                    {
                        logger.LogInformation("    Material '{MatName}' [{MatIndex}]: Found KHR_materials_pbrSpecularGlossiness.", matName, matIdx);
                        JsonObject pbrMetallicRoughness = mat["pbrMetallicRoughness"]?.AsObject() ?? new JsonObject();

                        if (specGloss.TryGetPropertyValue("diffuseTexture", out JsonNode? diffTexInfoNode) && diffTexInfoNode != null)
                        {
                            int? texIndex = diffTexInfoNode["index"]?.GetValue<int>();
                            int? sourceImgIndex = GetSourceImageIndex(textures, texIndex);
                            if (sourceImgIndex.HasValue && updatedImageUriMap.ContainsKey(sourceImgIndex.Value))
                            {
                                pbrMetallicRoughness["baseColorTexture"] = diffTexInfoNode.DeepClone();
                                logger.LogDebug("      Mapped diffuseTexture -> baseColorTexture");
                                matModified = true;
                            }
                        }
                        
                        if (specGloss.TryGetPropertyValue("diffuseFactor", out JsonNode? diffFactorNode) && diffFactorNode != null)
                        {
                            pbrMetallicRoughness["baseColorFactor"] = diffFactorNode.DeepClone();
                            matModified = true;
                        }

                        if (specGloss.TryGetPropertyValue("specularGlossinessTexture", out JsonNode? sgTexInfoNode) && sgTexInfoNode != null)
                        {
                            int? texIndex = sgTexInfoNode["index"]?.GetValue<int>();
                            int? sourceImgIndex = GetSourceImageIndex(textures, texIndex);
                            if (sourceImgIndex.HasValue && updatedImageUriMap.ContainsKey(sourceImgIndex.Value))
                            {
                                pbrMetallicRoughness["metallicRoughnessTexture"] = sgTexInfoNode.DeepClone();
                                pbrMetallicRoughness["metallicFactor"] = 0.0;
                                pbrMetallicRoughness["roughnessFactor"] = 1.0;
                                logger.LogDebug("      Mapped specularGlossinessTexture -> metallicRoughnessTexture (using generated _roughness map)");
                                matModified = true;
                            }
                        }

                        if (!matModified && mat.TryGetPropertyValue("normalTexture", out JsonNode? normTexInfoNode) && normTexInfoNode != null)
                        {
                            int? texIndex = normTexInfoNode["index"]?.GetValue<int>();
                            int? sourceImgIndex = GetSourceImageIndex(textures, texIndex);
                            if (sourceImgIndex.HasValue && updatedImageUriMap.ContainsKey(sourceImgIndex.Value))
                            {
                                matModified = true;
                                logger.LogDebug("      Normal map URI was updated.");
                            }
                        }

                        if (matModified)
                        {
                            mat["pbrMetallicRoughness"] = pbrMetallicRoughness;
                            extensions.Remove("KHR_materials_pbrSpecularGlossiness");
                            specGlossExtensionRemoved = true;
                            if (extensions.Count == 0) mat.Remove("extensions");
                            logger.LogInformation("    Material '{MatName}' [{MatIndex}]: Converted to PBR MetallicRoughness.", matName, matIdx);
                        }
                    }
                    else
                    {
                        JsonObject? pbr = mat["pbrMetallicRoughness"]?.AsObject();
                        JsonNode?[] textureNodes = {
                            mat["normalTexture"], 
                            mat["occlusionTexture"], 
                            mat["emissiveTexture"],
                            pbr?["baseColorTexture"], 
                            pbr?["metallicRoughnessTexture"]
                        };

                        foreach(var texInfoNode in textureNodes)
                        {
                            if (texInfoNode == null) continue;
                            
                            int? texIndex = texInfoNode["index"]?.GetValue<int>();
                            int? sourceImgIndex = GetSourceImageIndex(textures, texIndex);
                            if (sourceImgIndex.HasValue && updatedImageUriMap.ContainsKey(sourceImgIndex.Value))
                            {
                                matModified = true;
                                logger.LogDebug("    Material '{MatName}' [{MatIndex}]: Referenced image URI was updated.", matName, matIdx);
                                break;
                            }
                        }
                    }
                    
                    if(matModified) fileUpdated = true;
                }
            }

            if (specGlossExtensionRemoved)
            {
                RemoveExtension(rootNode, "extensionsUsed", "KHR_materials_pbrSpecularGlossiness");
                RemoveExtension(rootNode, "extensionsRequired", "KHR_materials_pbrSpecularGlossiness");
            }

            if (fileUpdated)
            {
                // Save updated GLTF file with "_converted" suffix
                string originalFileName = Path.GetFileName(filePath);
                string newFileName = Utils.InsertSuffix(originalFileName, "_converted");
                string outputGltfPath = Path.Combine(Path.GetDirectoryName(filePath) ?? string.Empty, newFileName);

                logger.LogInformation("  Saving updated GLTF file to: {OutputGltfPath}", outputGltfPath);
                try
                {
                    var writeOptions = new JsonSerializerOptions
                    {
                        WriteIndented = true,
                        Encoder = System.Text.Encodings.Web.JavaScriptEncoder.Create(System.Text.Unicode.UnicodeRanges.All)
                    };

                    // Ensure exclusive file access
                    using var fileStream = new FileStream(
                        outputGltfPath,
                        FileMode.Create,
                        FileAccess.Write,
                        FileShare.None, // Prevent other processes from accessing the file
                        bufferSize: 4096,
                        useAsync: true);
                    // Write to file stream instead of directly to file
                    var jsonString = rootNode.ToJsonString(writeOptions);
                    var bytes = System.Text.Encoding.UTF8.GetBytes(jsonString);
                    await fileStream.WriteAsync(bytes, cancellationToken);
                    await fileStream.FlushAsync(cancellationToken);
                }
                catch (IOException ioEx) when (ioEx.Message.Contains("being used by another process"))
                {
                    // If file is in use, try an alternative name
                    string altFileName = Utils.InsertSuffix(originalFileName, $"_converted_{Guid.NewGuid().ToString("N").Substring(0, 8)}");
                    string altOutputPath = Path.Combine(Path.GetDirectoryName(filePath) ?? string.Empty, altFileName);
                    
                    logger.LogWarning("  Original output file is in use. Trying alternative name: {AltOutputPath}", altOutputPath);
                    
                    try
                    {
                        var writeOptions = new JsonSerializerOptions
                        {
                            WriteIndented = true,
                            Encoder = System.Text.Encodings.Web.JavaScriptEncoder.Create(System.Text.Unicode.UnicodeRanges.All)
                        };
                        
                        await File.WriteAllTextAsync(altOutputPath, rootNode.ToJsonString(writeOptions), cancellationToken);
                    }
                    catch (Exception altEx)
                    {
                        logger.LogError(altEx, "  Error writing to alternative output path: {AltOutputPath}", altOutputPath);
                    }
                }
                catch (OperationCanceledException)
                {
                    logger.LogWarning("  Writing updated GLTF file cancelled: {OutputGltfPath}", outputGltfPath);
                    throw;
                }
                catch (Exception ex)
                {
                    logger.LogError(ex, "  Error writing updated GLTF: {OutputGltfPath}", outputGltfPath);
                }
            }
            else
            {
                logger.LogInformation("  No relevant texture URIs or material structures needed updating in {FileName}.", Path.GetFileName(filePath));
            }
        }
        catch (JsonException jsonEx)
        {
            logger.LogError(jsonEx, "  Error processing GLTF file {FilePath}", filePath);
        }
        catch (OperationCanceledException)
        {
            logger.LogWarning("GLTF processing cancelled for file: {FilePath}", filePath);
            throw;
        }
        catch (Exception ex)
        {
            logger.LogError(ex, "  Unexpected error processing GLTF file {FilePath}", filePath);
        }
        
        return fileUpdated;
    }

    private static int? GetSourceImageIndex(JsonArray? textures, int? textureIndex)
    {
        if (textures == null || !textureIndex.HasValue || textureIndex.Value < 0 || textureIndex.Value >= textures.Count)
            return null;
        return textures[textureIndex.Value]?["source"]?.GetValue<int>();
    }

    private static void RemoveExtension(JsonNode root, string arrayName, string extensionToRemove)
    {
        if (root[arrayName] is JsonArray extensionsArray)
        {
            JsonNode? toRemove = extensionsArray.FirstOrDefault(node => 
                node?.GetValue<string>() == extensionToRemove);
                
            if (toRemove != null)
            {
                extensionsArray.Remove(toRemove);
                if (extensionsArray.Count == 0 && root is JsonObject rootObj)
                {
                    rootObj.Remove(arrayName);
                }
            }
        }
    }
}

File: src/tarkov_texture_converter/gltf_utils.py
import os
import json
import logging
from typing import Dict

from .constants import TextureType
from .processor import TextureProcessor
from .utils import insert_suffix

logger = logging.getLogger(__name__)

def update_gltf_files(input_folder: str, output_folder_abs_path: str):
    """Scans and updates GLTF files for Tarkin mode."""
    if not os.path.isdir(input_folder):
        logger.error(f"GLTF Update: Input folder '{input_folder}' not found.")
        return

    output_folder_basename = os.path.basename(output_folder_abs_path)
    logger.info(f"Scanning for GLTF files in '{input_folder}' relative to '{output_folder_basename}'...")
    gltf_found_count, updated_gltf_count = 0, 0

    try:
        for entry in os.scandir(input_folder):
            if entry.is_file() and entry.name.lower().endswith(".gltf"):
                 gltf_found_count += 1
                 gltf_path = entry.path
                 logger.info(f"Processing GLTF file: {gltf_path}")
                 try:
                     with open(gltf_path, "r", encoding='utf-8') as f: gltf_data = json.load(f)
                 except Exception as e:
                     logger.error(f"  Error loading/decoding {gltf_path}: {e}")
                     continue

                 materials = gltf_data.get("materials", [])
                 images = gltf_data.get("images", [])
                 textures = gltf_data.get("textures", [])
                 file_updated = False
                 image_uri_map: Dict[int, str] = {}

                 for img_index, img in enumerate(images):
                     old_uri = img.get("uri", "")
                     if not old_uri or old_uri.startswith("data:"): continue
                     old_uri_norm = old_uri.replace("\\", "/")
                     old_filename = os.path.basename(old_uri_norm)
                     texture_type = TextureProcessor._get_texture_type(old_filename, tarkin_mode=True)
                     new_uri_base = ""
                     if texture_type == TextureType.DIFFUSE: new_uri_base = insert_suffix(old_filename, "_color") + ".png"
                     elif texture_type == TextureType.SPECGLOS: new_uri_base = insert_suffix(old_filename, "_roughness") + ".png"
                     elif texture_type == TextureType.NORMAL: new_uri_base = insert_suffix(old_filename, "_converted") + ".png"

                     if new_uri_base:
                         new_uri = os.path.join(output_folder_basename, new_uri_base).replace("\\", "/")
                         if img.get("uri") != new_uri:
                             logger.info(f"    Image [{img_index}]: Updating URI '{img.get('uri')}' -> '{new_uri}'")
                             img["uri"] = new_uri
                             image_uri_map[img_index] = new_uri
                             file_updated = True

                 used_extensions = set(gltf_data.get("extensionsUsed", []))
                 required_extensions = set(gltf_data.get("extensionsRequired", []))
                 removed_specgloss_extension = False

                 for mat_index, mat in enumerate(materials):
                     mat_name = mat.get("name", f"Material_{mat_index}")
                     pbr_metallic_roughness = mat.get("pbrMetallicRoughness", {})
                     extensions = mat.get("extensions", {})
                     spec_gloss = extensions.get("KHR_materials_pbrSpecularGlossiness")
                     mat_updated_internally = False

                     if spec_gloss:
                         # Base Color
                         tex_info = spec_gloss.get("diffuseTexture")
                         tex_index = tex_info.get("index") if tex_info else None
                         if tex_index is not None and 0 <= tex_index < len(textures):
                             source_img_index = textures[tex_index].get("source")
                             if source_img_index in image_uri_map:
                                 pbr_metallic_roughness["baseColorTexture"] = tex_info
                                 mat_updated_internally = True
                         if "diffuseFactor" in spec_gloss:
                            pbr_metallic_roughness["baseColorFactor"] = spec_gloss["diffuseFactor"]
                            mat_updated_internally = True
                         # Metallic/Roughness
                         tex_info = spec_gloss.get("specularGlossinessTexture")
                         tex_index = tex_info.get("index") if tex_info else None
                         if tex_index is not None and 0 <= tex_index < len(textures):
                             source_img_index = textures[tex_index].get("source")
                             if source_img_index in image_uri_map:
                                 pbr_metallic_roughness["metallicRoughnessTexture"] = tex_info
                                 pbr_metallic_roughness.setdefault("metallicFactor", 0.0)
                                 pbr_metallic_roughness.setdefault("roughnessFactor", 1.0)
                                 mat_updated_internally = True
                         # Normal Check
                         tex_info = mat.get("normalTexture")
                         tex_index = tex_info.get("index") if tex_info else None
                         if tex_index is not None and 0 <= tex_index < len(textures):
                              source_img_index = textures[tex_index].get("source")
                              if source_img_index in image_uri_map: mat_updated_internally = True

                         if mat_updated_internally:
                             mat["pbrMetallicRoughness"] = pbr_metallic_roughness
                             extensions.pop("KHR_materials_pbrSpecularGlossiness")
                             if not extensions: mat.pop("extensions", None)
                             removed_specgloss_extension = True
                             logger.info(f"    Material '{mat_name}' [{mat_index}]: Converted to PBR MetRough.")
                     else: # Standard PBR check
                         keys_to_check = ["baseColorTexture", "metallicRoughnessTexture", "normalTexture", "occlusionTexture", "emissiveTexture"]
                         locs = [mat, pbr_metallic_roughness]
                         updated = False
                         for key in keys_to_check:
                              for loc in locs:
                                   tex_info = loc.get(key)
                                   if tex_info and isinstance(tex_info, dict):
                                       idx = tex_info.get("index")
                                       if idx is not None and 0 <= idx < len(textures):
                                            src_idx = textures[idx].get("source")
                                            if src_idx in image_uri_map: mat_updated_internally = True; updated=True; break
                              if updated: break

                     if mat_updated_internally: file_updated = True

                 if file_updated:
                     if removed_specgloss_extension:
                         if "KHR_materials_pbrSpecularGlossiness" in used_extensions:
                             used_extensions.remove("KHR_materials_pbrSpecularGlossiness")
                             if used_extensions: gltf_data["extensionsUsed"] = sorted(list(used_extensions))
                             else: gltf_data.pop("extensionsUsed", None)
                         if "KHR_materials_pbrSpecularGlossiness" in required_extensions:
                              required_extensions.remove("KHR_materials_pbrSpecularGlossiness")
                              if required_extensions: gltf_data["extensionsRequired"] = sorted(list(required_extensions))
                              else: gltf_data.pop("extensionsRequired", None)
                     output_gltf_path = gltf_path
                     logger.info(f"  Saving updated GLTF file to: {output_gltf_path}")
                     try:
                         with open(output_gltf_path, "w", encoding='utf-8') as f:
                             json.dump(gltf_data, f, indent=2, ensure_ascii=False, separators=(',', ': '))
                         updated_gltf_count += 1
                     except Exception as e: logger.error(f"  Error writing updated GLTF: {e}")
                 else: logger.info(f"  No relevant changes in {entry.name}.")

    except OSError as e: logger.error(f"GLTF Update: Error scanning {input_folder}: {e}")
    except Exception as e: logger.error(f"GLTF Update: Unexpected error: {e}", exc_info=True)

    if gltf_found_count == 0: logger.info("GLTF Update: No .gltf files found.")
    else: logger.info(f"GLTF Update: Processed {gltf_found_count} GLTF files, updated {updated_gltf_count}.")